File: packages/snow_draw_rust_engine/src/draw/elements/types/filter/filter_task_encoder.rs
#![allow(dead_code)]

use std::sync::Arc;

use crate::draw::elements::core::typed_element_render_task_encoder::{
    ElementState, RenderTaskList, TypedElementRenderTaskEncoder,
};
use crate::draw::types::edit_context::TextMetricsService;

use super::filter_data::FilterData;

/// Element-level render task for filter payloads.
///
/// Mirrors Dart `FilterRenderTask(element: ..., data: ..., localeTag: ...)`.
#[derive(Debug, Clone)]
pub struct FilterRenderTask {
    pub element: ElementState,
    pub data: FilterData,
    pub locale_tag: Option<String>,
}

impl FilterRenderTask {
    /// Creates a new filter render task snapshot.
    pub fn new(element: ElementState, data: FilterData, locale_tag: Option<String>) -> Self {
        Self {
            element,
            data,
            locale_tag,
        }
    }
}

/// Encodes filter elements into high-level render tasks.
///
/// Translation of Dart
/// `FilterTaskEncoder extends TypedElementRenderTaskEncoder<FilterData>`.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct FilterTaskEncoder;

impl FilterTaskEncoder {
    /// Creates a filter task encoder.
    pub const fn new() -> Self {
        Self
    }
}

impl TypedElementRenderTaskEncoder<FilterData> for FilterTaskEncoder {
    fn encode_typed_tasks(
        &self,
        element: &ElementState,
        data: &FilterData,
        locale_tag: Option<&str>,
        _text_metrics_service: Option<Arc<dyn TextMetricsService>>,
    ) -> RenderTaskList {
        vec![Box::new(FilterRenderTask::new(
            element.clone(),
            data.clone(),
            locale_tag.map(str::to_owned),
        ))]
    }
}

File: packages/snow_draw_engine/lib/draw/elements/types/filter/filter_task_encoder.dart
import '../../../models/element_state.dart';
import '../../../render/tasks/render_tasks.dart';
import '../../../services/text/text_metrics_service.dart';
import '../../core/typed_element_render_task_encoder.dart';
import 'filter_data.dart';

/// Encodes filter elements into high-level render tasks.
final class FilterTaskEncoder
    extends TypedElementRenderTaskEncoder<FilterData> {
  /// Creates a filter task encoder.
  const FilterTaskEncoder();

  @override
  List<RenderTask> encodeTypedTasks({
    required ElementState element,
    required FilterData data,
    String? localeTag,
    TextMetricsService? textMetricsService,
  }) => <RenderTask>[
    FilterRenderTask(element: element, data: data, localeTag: localeTag),
  ];
}
